Simplenote/Controllers/PublishController.swift

```diff
@@ -0,0 +1,101 @@
+import Foundation
+
+@objc
+class PublishController: NSObject {
+    private var callbackMap: PublishMap
+    private let timerFactory: TimerFactory
+
+    init(timerFactory: TimerFactory = TimerFactory(), callbackMap: PublishMap = PublishMap()) {
+        self.timerFactory = timerFactory
+        self.callbackMap = callbackMap
+    }
+
+    func updatePublishState(for note: Note, to published: Bool, completion: @escaping (PublishState) -> Void) {
+        if note.published == published {
+            return
+        }
+
+        let wrapper = PublishListenWrapper(note: note, block: completion)
+        callbackMap.container[note.simperiumKey] = wrapper
+
+        SPTracker.trackEditorNotePublishEnabled(published)
+        changePublishState(for: note, to: published)
+
+        published ? update(wrapper, to: .publishing) : update(wrapper, to: .unpublishing)
+    }
+
+    @objc(didReceiveUpdateFromSimperiumForKey:)
+    func didReceiveUpdateFromSimperium(for key: NSString) {
+        guard let wrapper = callbackMap.container[key as String] else {
+            return
+        }
+
+        if wrapper.note.published && wrapper.note.publishURL != nil {
+            update(wrapper, to: .published)
+            return
+        }
+
+        if !wrapper.note.published {
+            update(wrapper, to: .unpublished)
+
+            return
+        }
```

Comment:
How about if the `wrapper` encapsulates the logic to calculate the actual `PublishedState` ?

(That would simplify this API a lot!)